### PR TITLE
Introduce tabs for notebook views

### PR DESCRIPTION
**Pull request by dhgrajeda**
Cards have been restructured to contain tabs instead of cells directly.
Tabs then contain the cards that are displayed. Cards may contain one or
more cards.

The panel component has been reused to implement the cells. There is
still some cleanup left to do of unused functions, but I want to reduce
the size of the PR as much as possible.

image

### CODE REVIEW

**Comment by Charles-Gagnon** [comment]
Please fix the test issues

**Comment by corivera** [comment]
Could we get a screenshot of what this looks like?

**Comment by serveramigo1** [comment]
LUIS

**Review by serveramigo1 — COMMENTED** [review]
LUIS

File: src/sql/workbench/contrib/notebook/browser/notebookViews/notebookViewsCard.component.html [hunk 1]
```diff
@@ -6,27 +6,37 @@
...
-					<text-cell-component *ngIf="cell.cellType === 'markdown'" [cellModel]="cell" [model]="model" [activeCellId]="activeCellId">
-					</text-cell-component>
+					<div #actionbar [style.display]="showActionBar ? 'block' : 'none'" class="actionbar"></div>
+					<panel (onTabChange)="handleTabChange($event)">
+						<tab [visibilityType]="'visibility'" *ngFor="let tab of tabs" [title]="tab.title" class="fullsize" [identifier]="tab.id" [type]="'tab'">
+							<ng-template>
+								<ng-container>
+									<div #container class="card-container">
+										<div *ngIf="tab.cellModel !== undefined">
+											<views-code-cell-component *ngIf="tab.cellModel.cellType === 'code'" [cellModel]="tab.cellModel" [model]="model" [activeCellId]="tab.cellModel.id + 000" [visible]="visible">
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookViews/notebookViewsCard.component.html:30]: What's this "+ 000" for?

File: src/sql/workbench/services/notebook/browser/notebookViews/notebookViewModel.ts [hunk 2]
```diff
@@ -31,44 +64,66 @@ export class NotebookViewModel implements INotebookView {
...
+			this._cards.push(card);
+		});
+	}
+
+	protected createTab(cell: ICellModel, card: INotebookViewCard): void {
+		if (card === undefined) {
+			throw new Error('A card must be specified to create a tab');
 		}
+
+		const newTab: ViewsTabConfig = { title: 'Untitled', id: generateUuid(), group: card.guid, cell: { guid: cell.cellGuid } };
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/notebookViews/notebookViewModel.ts:125]: Should untitled be localized?

File: src/sql/workbench/services/notebook/browser/notebookViews/notebookViewModel.ts [hunk 3]
```diff
@@ -111,30 +171,63 @@ export class NotebookViewModel implements INotebookView {
 		return this._notebookViews.notebook.cells.find(cell => cell.cellGuid === guid);
 	}
 
-	public updateCell(cell: ICellModel, currentView: INotebookView, cellData: INotebookViewCell, override: boolean = false) {
+	public updateCell(cell: ICellModel, currentView: INotebookView, cellData: INotebookViewCard, override: boolean = false) {
 		if (!this.cellInitialized(cell)) {
-			this.initializeCell(cell, 0);
+			//this.initializeCell(cell, 0);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/notebookViews/notebookViewModel.ts:176]: Remove?

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by Charles-Gagnon** [comment]
@dhgrajeda Please also always include screenshots/gifs when making any UI changes

**Comment by dhgrajeda** [comment]
Added a screenshot to the description

File: src/sql/workbench/services/notebook/browser/notebookViews/notebookViewsExtension.ts [hunk 4]
```diff
@@ -127,7 +127,7 @@ export class NotebookViewsExtension extends NotebookExtension<INotebookViewMetad
 		return i <= this.maxNameIterationAttempts ? name : generateUuid();
 	}
 
-	public updateCell(cell: ICellModel, currentView: INotebookView, cellData: INotebookViewCell, override: boolean = false) {
+	public updateCell(cell: ICellModel, currentView: INotebookView, cellData: INotebookViewCard, override: boolean = false) {
```
[the remark this answers is not in the document]
Comment (dhgrajeda) [reply src/sql/workbench/services/notebook/browser/notebookViews/notebookViewsExtension.ts:134]: You're right. I did not rename all of the functions/variables/types on purpose to reduce the pull request size. I will be cleaning this up in a separate pull request.

File: src/sql/workbench/contrib/notebook/browser/notebookViews/notebookViewsCard.component.html (hunk 1 above)
Comment (dhgrajeda) [reply src/sql/workbench/contrib/notebook/browser/notebookViews/notebookViewsCard.component.html]: This was something I had to clean up. I've removed the zeros.

**Pull request merged**